service/address_map.hh

```diff
@@ -181,102 +213,82 @@ class address_map_t : public peering_sharded_service<address_map_t<Clock>> {
-            bool update_if_exists) {
-        replicate([id, generation_number, ip_addr, update_if_exists] (address_map_t& self) {
-            self.handle_add_or_update_entry(id, generation_number, ip_addr, update_if_exists);
-        });
+    static void apply(cluster_mutation& dst, const cluster_mutation& src) {
+        for (auto&& e : src) {
+            apply(dst, e);
+        }
     }
 
-    void handle_set_nonexpiring(const locator::host_id& id) {
-        auto [it, _] = _map.try_emplace(id, timestamped_entry{gms::generation_type{}, std::nullopt});
-        auto& entry = it->second;
-
-        if (entry.expiring()) {
-            entry._lru_entry = nullptr;
-        }
+    void on_replication_failed(std::exception_ptr e) {
+        rslog.warn("address map replication failed: {}", e);
     }
 
-    void handle_set_expiring(const locator::host_id& id) {
-        auto it = _map.find(id);
-        if (it == _map.end()) {
-            return;
-        }
-        auto& entry = it->second;
-        if (entry.expiring()) {
+    // Strong exception guarantees: no side effects on exception.
+    void apply_locally(const host_mutation& mf, bool update_if_exists = true) {
+        auto&& [id, m]  = mf;
+        if (!m) {
             return;
         }
-        add_expiring_entry(it->first, entry);
-    }
 
-    void handle_add_or_update_entry(const locator::host_id& id,
-            gms::generation_type generation_number, const gms::inet_address& ip_addr,
-            bool update_if_exists) {
-        auto [it, emplaced] = _map.try_emplace(id, timestamped_entry{generation_number, ip_addr});
+        auto [it, emplaced] = _map.try_emplace(id, timestamped_entry{m});
         auto& entry = it->second;
+
         if (emplaced) {
-            add_expiring_entry(it->first, entry);
-        } else if ((update_if_exists && generation_number >= entry._generation_number) || !entry._addr) {
-            entry._addr = ip_addr;
-            entry._generation_number = generation_number;
+            if (!m.expiring || *m.expiring) {
+                try {
+                    add_expiring_entry(it->first, entry);
+                } catch (...) {
+                    _map.erase(it);
+                    throw;
+                }
+            }
+        } else if (update_if_exists) {
+            if (m.expiring) {
+                if (*m.expiring) {
+                    // Do first, so that we don't leave side effects on exception
+                    if (!entry.expiring()) {
+                        add_expiring_entry(it->first, entry);
+                    }
+                } else {
+                    entry._lru_entry = nullptr;
+                }
+            }
+            entry._m.apply(m);
             if (entry.expiring()) {
-                entry._lru_entry->touch(); // Re-insert in the front of _expiring_list
+                entry._lru_entry->touch();
             }
         }
     }
 
+    void apply_locally(const cluster_mutation& m) {
```

Comment:
I'd add a comment that this entry point doesn't provide strong exception guarantees so that
some host_mutations in `m` may be applied if an exception is thrown, and they are not rolled back.